packages/socket/patches/socket.io-parser+4.0.4.patch

```diff
@@ -146,7 +146,11 @@ index 4b7c234..95469f7 100644
 +    known.push(obj)
      if (Array.isArray(obj)) {
          for (let i = 0, l = obj.length; i < l; i++) {
-             if (hasBinary(obj[i])) {
+-            if (hasBinary(obj[i])) {
++            if (hasBinary(obj[i], known)) {
+                 return true;
+             }
+         }
 @@ -43,10 +49,10 @@ function hasBinary(obj, toJSON) {
      if (obj.toJSON &&
```

Comment:
Random observation: this `if` condition is always false, because it assumes the `hasBinary` funciton can be called with only one argument. It was possible in the original code (optional `toJson` argument) , but not possible with the patched version

Reply:
I don't think so.

The full code is like below:

```js
function hasBinary(obj, known = []) {
    if (!obj || typeof obj !== "object") {
        return false;
    }
    if (known.includes(obj)) {
        return false
    }
    known.push(obj)
    if (Array.isArray(obj)) {
        for (let i = 0, l = obj.length; i < l; i++) {
            if (hasBinary(obj[i], known)) {
                return true;
            }
        }
        return false;
    }
    if (isBinary(obj)) {
        return true;
    }
    if (obj.toJSON &&
        typeof obj.toJSON === "function" &&
        arguments.length === 1) {
        return hasBinary(obj.toJSON(), known);
    }
    for (const key in obj) {
        if (Object.prototype.hasOwnProperty.call(obj, key) && hasBinary(obj[key], known)) {
            return true;
        }
    }
    return false;
}
```

When we pass `[1, {}]` as `obj`. `hasBinary(1, [[1, {}]])` returns `false` but `hasBinary({}, [[1, {}], 1])` returns `true`.

Reply:
Sorry I probably was not very clear. I specifically meant the `if` condition that handles a very special use case, when an object has `toJSON` method to be called.
This `if` condition checks that `aruments.length === 1` - which is never the case in the patched version (both your examples have `arguments.length === 2` because of the `known` argument).

Reply:
I checked [the original code here](https://github.com/socketio/socket.io-parser/blob/75530b4dcdced9138d51a0b7b782c8f89cdf3899/lib/is-binary.ts#L33-L66) and it seems that the second arg, `toJson` is a dead arg (it doesn't affect the flow of the code) and it might exist because of the backward compatibility. 

```ts
export function hasBinary(obj: any, toJSON?: boolean) {
  if (!obj || typeof obj !== "object") {
    return false;
  }

  if (Array.isArray(obj)) {
    for (let i = 0, l = obj.length; i < l; i++) {
      if (hasBinary(obj[i])) {
        return true;
      }
    }
    return false;
  }

  if (isBinary(obj)) {
    return true;
  }

  if (
    obj.toJSON &&
    typeof obj.toJSON === "function" &&
    arguments.length === 1
  ) {
    return hasBinary(obj.toJSON(), true);
  }

  for (const key in obj) {
    if (Object.prototype.hasOwnProperty.call(obj, key) && hasBinary(obj[key])) {
      return true;
    }
  }

  return false;
}
```

As for an object which has `toJSON` function, it's properly handled with the code below:

```js
    if (obj.toJSON &&
        typeof obj.toJSON === "function" &&
        arguments.length === 1) {
        return hasBinary(obj.toJSON(), known);
    }
```

Reply:
Sorry. Forget the old comment. I misunderstood your comment again. 

I believe this can be fixed by removing `arguments.length === 1`. Maybe I need to create a new test case.

Reply:
@amakhrov I opened the issue at #17988.

Reply:
awesome!